my/coding/commits.py

```diff
@@ -139,35 +143,43 @@ def repos():
     return git_repos_in(config.roots)
 
 
-def _hashf(_repos: List[Path]):
-    # TODO maybe use smth from git library? ugh..
-    res = []
-    for r in _repos:
-        # TODO just use anything except index? ugh.
-        for pp in {
-                '.git/FETCH_HEAD',
-                '.git/HEAD',
-                'FETCH_HEAD', # bare
-                'HEAD', # bare
-        }:
-            ff = r / pp
-            if ff.exists():
-                updated = ff.stat().st_mtime
-                break
-        else:
-            raise RuntimeError(r)
-        res.append((r, updated))
-    return res
-
-# TODO per-repo cache?
-# TODO set default cache path?
-# TODO got similar issue as in photos with a helper method.. figure it out
-@mcachew(hashf=_hashf, logger=log)
-def _commits(_repos) -> Iterator[Commit]:
+# returns modification time for an index to use as hash function
+def _repo_depends_on(_repo: Path) -> int:
+    for pp in {
+        ".git/FETCH_HEAD",
+        ".git/HEAD",
+        "FETCH_HEAD",  # bare
+        "HEAD",  # bare
+    }:
+        ff = _repo / pp
+        if ff.exists():
+            return int(ff.stat().st_mtime)
+    else:
+        raise RuntimeError(f"Could not find a FETCH_HEAD/HEAD file in {_repo}")
+
+
+def _commits(_repos: List[Path]) -> Iterator[Commit]:
     for r in _repos:
-        log.info('processing %s', r)
-        yield from repo_commits(r)
+        yield from _cached_commits(r)
+
+_allowed_letters: str = string.ascii_letters + string.digits
+
+
+def _cached_commits_path(p: Path) -> str:
+    # compute a reduced simple filepath using the absolute path of the repo
+    simple_path = ''.join(filter(lambda c: c in _allowed_letters, str(p.absolute())))
+    return cache_dir() / simple_path / '_cached_commits'
```

Comment:
Maybe worth having a subdirectory as well (i.e. `cache_dir() / 'commits'), otherwise it pollutes the top level cache dir. Not a problem, but makes it a bit more confusing

Reply:
Yeah, I could reverse it so that `_cached_commits` is the top-level subdirectory and each repo gets a folder in there.

Maybe I should note that this 'simplified path' is something [I used to have in core](https://github.com/seanbreckenridge/HPI/blob/6e2262d66b332a83099e3740aa3559204427e95b/my/core/file.py#L76-L84), because it was also useful for `my.google`, to quickly be able to create a [`cache_path`](https://github.com/seanbreckenridge/HPI/blob/6e2262d66b332a83099e3740aa3559204427e95b/my/coding/commits.py#L185-L192) based on a simplified version of the filepath. I don't know if you've run into this issue before but I often did when parsing large GDPR exports. Maybe its worth to have a helper for this?

Reply:
> Yeah, I could reverse it so that _cached_commits is the top-level subdirectory and each repo gets a folder in there.
I'm about to submit a PR with some more cleanup, was gonna share with you!

> Maybe its worth to have a helper for this?
Yep, sure!